python/beeai_framework/adapters/a2a/agents/agent.py

```diff
@@ -0,0 +1,275 @@
+    ) -> None:
+        super().__init__()
+        if agent_card:
+            self._name: str = agent_card.name
+            self._url: str = agent_card.url
+        elif url:
+            self._name = f"agent_{url.split(':')[-1]}"
+            self._url = url
+        else:
+            raise ValueError("Either url or agent_card must be provided.")
+        if not memory.is_empty():
+            raise ValueError("Memory must be empty before setting.")
+        self._memory: BaseMemory = memory
+        self._agent_card: a2a_types.AgentCard | None = agent_card
+        self._context_id: str | None = None
+        self._task_id: str | None = None
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    def run(
+        self,
+        input: str | AnyMessage | a2a_types.Message,
+        *,
+        context_id: str | None = None,
+        task_id: str | None = None,
+        signal: AbortSignal | None = None,
+    ) -> Run[A2AAgentRunOutput]:
+        self._context_id = context_id or self._context_id
+        self._task_id = task_id or self._task_id
+
+        async def handler(context: RunContext) -> A2AAgentRunOutput:
+            async with httpx.AsyncClient() as httpx_client:
+                client: a2a_client.A2AClient = await a2a_client.A2AClient.get_client_from_agent_card_url(
+                    httpx_client, self._url
+                )
+
+                streaming_request: a2a_types.SendStreamingMessageRequest = a2a_types.SendStreamingMessageRequest(
+                    params=a2a_types.MessageSendParams(message=self._convert_to_a2a_message(input))
+                )
+
+                last_event_with_data = None
+                stream_response: AsyncGenerator[a2a_types.SendStreamingMessageResponse] = client.send_message_streaming(
+                    streaming_request
+                )
+                async for event in stream_response:
+                    if last_event_with_data is None or has_content(event):
+                        last_event_with_data = event
+                    # emit all events as updates
+                    await context.emitter.emit(
+                        "update", A2AAgentUpdateEvent(value=event.model_dump(mode="json", exclude_none=True))
+                    )
+
+                # check if we received a any response
+                if last_event_with_data is None:
+                    raise AgentError("No result received from agent.")
+
+                # process error
+                if isinstance(last_event_with_data.root, a2a_types.JSONRPCErrorResponse):
+                    await context.emitter.emit(
+                        "error",
+                        A2AAgentErrorEvent(message=last_event_with_data.root.error.message or "Unknown error"),
+                    )
+                    raise AgentError(last_event_with_data.root.error.message or "Unknown error")
```

Comment:
Maybe I would leverage the `context` attribute to pass some additional data, like ``last_event_with_data